src/Workspaces/Core/Portable/Utilities/AbstractSpeculationAnalyzer.cs

```diff
@@ -199,6 +199,24 @@ private void EnsureSpeculativeSemanticModel()
 
         #region Semantic comparison helpers
 
+        private bool ReplacementIntroducesErrorType(TExpressionSyntax originalExpression, TExpressionSyntax newExpression)
+        {
+            Debug.Assert(originalExpression != null);
+            Debug.Assert(this.SemanticRootOfOriginalExpression.DescendantNodesAndSelf().Contains(originalExpression));
+            Debug.Assert(newExpression != null);
+            Debug.Assert(this.SemanticRootOfReplacedExpression.DescendantNodesAndSelf().Contains(newExpression));
```

Comment:
[minor] These checks seem to be repeated across bunch of methods below. Consider pulling out into a separate method?